platform/projectModel-api/src/com/intellij/openapi/roots/DependencyScope.java

```diff
@@ -97,6 +97,23 @@ public boolean isForTestRuntime() {
     return myForTestRuntime;
   }
 
+  public @NotNull DependencyScope including(@NotNull DependencyScope other) {
+    if (this == other || this.includes(other)) return this;
+    if (other.includes(this)) return other;
+    // At this point, we know that `this` and `other` are different, neither is COMPILE, and they're not PROVIDED+TEST either.
+    // The remaining cases are:
+    // - PROVIDED+RUNTIME = COMPILE
+    // - TEST+RUNTIME = COMPILE (broader than necessary, but there's no closer match)
+    return COMPILE;
+  }
+
+  public boolean includes(@NotNull DependencyScope other) {
```

Comment:
I renamed and refactored this into `coversUseCasesOf(DependencyScope)`, let me know if the logic's still not clear.

I could extract the checks for each use-case into separate variables, but that would break the current short-circuiting optimization (though the compiler could still potentially reintroduce the same optimization by throwing the variables away, I don't know).